core/src/test/scala/unit/kafka/admin/RackAwareTest.scala

```diff
@@ -0,0 +1,101 @@
+    val leaderCount: Map[Int, Int] = distribution.brokerLeaderCount
+    val partitionCount: Map[Int, Int] = distribution.brokerReplicasCount
+    val partitionRackMap: Map[Int, Seq[String]] = distribution.partitionRacks
+    val numReplicasPerBroker = numPartitions * replicationFactor / numBrokers
+    val leaderCountPerBroker = numPartitions / numBrokers
+    if (verifyRackAware) {
+      assertEquals(List.fill(numPartitions)(replicationFactor), partitionRackMap.values.toList.map(_.size))
+      assertEquals(List.fill(numPartitions)(replicationFactor), partitionRackMap.values.toList.map(_.toSet.size))
+    }
+    if (verifyLeaderDistribution) {
+      assertEquals(List.fill(numBrokers)(leaderCountPerBroker), leaderCount.values.toList)
+    }
+    if (verifyReplicasDistribution) {
+      assertEquals(List.fill(numBrokers)(numReplicasPerBroker), partitionCount.values.toList)
+    }
+  }
+
+  def getReplicaDistribution(assignment: scala.collection.Map[Int, Seq[Int]], brokerRackMapping: Map[Int, String]): ReplicaDistributions = {
+    // ensure leader distribution is even
+    val leaderCount: mutable.Map[Int, Int] = mutable.Map()
+    // ensure total partitions distribution is even
+    val partitionCount: mutable.Map[Int, Int] = mutable.Map()
+    // ensure replicas are assigned to unique racks
+    val partitionRackMap: mutable.Map[Int, List[String]] = mutable.Map()
+    assignment.foreach {
+      case (partitionId, replicaList) => {
+        val leader = replicaList.head
+        leaderCount(leader) = leaderCount.getOrElse(leader, 0) + 1
+        for (brokerId <- replicaList) {
+          partitionCount(brokerId) = partitionCount.getOrElse(brokerId, 0) + 1
+          partitionRackMap(partitionId) =  brokerRackMapping(brokerId) :: partitionRackMap.getOrElse(partitionId, List())
+        }
+      }
+    }
+    ReplicaDistributions(partitionRackMap, leaderCount, partitionCount)
+  }
+
+  def ensureSameNumberOfBrokersInRack(brokerList: Seq[Int], brokerRackMap: Map[Int, String]): Unit = {
+    val subRackInfo = brokerRackMap.filterKeys(brokerList.contains(_))
+    if (brokerRackMap.size > 0 && subRackInfo.size != brokerList.size) {
+      throw new AdminOperationException("Incomplete broker-rack mapping supplied for broker list " + brokerList)
+    }
+    val reverseMap = AdminUtils.getInverseMap(subRackInfo)
+    if (reverseMap.values.map(_.size).toSet.size > 1) {
+      throw new AdminOperationException("The number of brokers are not the same for all racks: %s".format(reverseMap))
+    }
+  }
+}
+
+case class ReplicaDistributions(partitionRacks: Map[Int, Seq[String]], brokerLeaderCount: Map[Int, Int], brokerReplicasCount: Map[Int, Int])
+
+class SimpleRackLocator(zkClient: ZkClient, props: Properties) extends RackLocator(zkClient, props) {
+  override def getRackInfo(): Map[Int, String] = {
+    val brokerRackMap = props.asScala.map {
+      case (key, value) => (key.toInt, value)
```

Comment:
something about the scala of this makes me want to say it should be an implicit, that is a much bigger topic and change so i would say maybe not introduce that now but here is one of a lot of places we could without losing readability or performance reduce code. Maybe even try it with this change as your converting type only in RacLocator